PocketKit/Sources/Sync/Space.swift

```diff
@@ -272,57 +324,112 @@ public class Space {
         )
     }
 
-    func makeImagesController() -> NSFetchedResultsController<Image> {
-        let request = Requests.fetchUndownloadedImages()
-        request.sortDescriptors = [NSSortDescriptor(key: "source.absoluteString", ascending: true)]
-        return NSFetchedResultsController(
-            fetchRequest: request,
-            managedObjectContext: backgroundContext,
-            sectionNameKeyPath: nil,
-            cacheName: nil
-        )
+    func batchDeleteOrphanedSlates() throws {
+        let fetchRequest: NSFetchRequest<NSFetchRequestResult> = Slate.fetchRequest()
+        fetchRequest.predicate = NSPredicate(format: "slateLineup = NULL")
+        let deleteRequest = NSBatchDeleteRequest(fetchRequest: fetchRequest)
+        _ = try backgroundContext.performAndWait {
+            try backgroundContext.execute(deleteRequest)
+        }
     }
 
-    func backgroundObject<T: NSManagedObject>(with id: NSManagedObjectID) -> T? {
-        backgroundContext.performAndWait {
-            backgroundContext.object(with: id) as? T
+    /// Updates a `SlateLineup` from the specified remote object, on a child background context
+    /// - Parameter remote: the specified remote object
+    func updateLineup(from remote: SlateLineup.RemoteSlateLineup) throws {
+        let context = makeChildBackgroundContext()
+
+        context.performAndWait { [weak self] in
+            guard let self else { return }
+
+            let lineup = (try? self.fetchSlateLineup(byRemoteID: remote.id, context: context)) ??
+            SlateLineup(context: context, remoteID: remote.id, expermimentID: remote.experimentId, requestID: remote.requestId)
+
+            lineup.update(from: remote, in: self, context: context)
+        }
+        // save the child context
+        try context.performAndWait {
+            guard context.hasChanges else {
+                return
+            }
+            try context.save()
```

Comment:
@Gio2018 I think we should do the batch deleting of orphans before we save. That would represent a full logical transaction. And ensure we save within this code path only once after we do everything we need to do to trigger a view update.

Reply:
fwiw I don't think batch deleting affects anything on a save, since its behavior is different, so i'm not convinced that order necessarily matters, but i understand what you're suggesting.

> [NSManagedObjectContext](https://developer.apple.com/documentation/coredata/nsmanagedobjectcontext) doesn’t automatically merge a request’s deletions because they happen at the SQL level. Subsequently, you must remove any deleted objects from memory after the request finishes. To determine the objects a request deletes, configure it to return the [NSManagedObjectID](https://developer.apple.com/documentation/coredata/nsmanagedobjectid) of each deleted object and use those identifiers to update your contexts, as the following example shows:

https://developer.apple.com/documentation/coredata/nsbatchdeleterequest

Reply:
Thanks for the comments @bassrock @dskuza ! I updated the code in favor of non-batch deletion of orphaned items, since we are not dealing with a lot of data here.